jiant/pytorch_transformers_interface/modules.py

```diff
@@ -0,0 +1,300 @@
+                    "model is frozen."
+                )
+            # TODO: if doing multiple target tasks, allow for multiple sets of
+            # scalars. See the ELMo implementation here:
+            # https://github.com/allenai/allennlp/blob/master/allennlp/modules/elmo.py#L115
+            assert len(parse_task_list_arg(args.target_tasks)) <= 1, (
+                "pytorch_transformers_embedding_mode='mix' only supports a single set of "
+                "scalars (but if you need this feature, see the TODO in "
+                "the code!)"
+            )
+            num_layers = self.model.config.num_hidden_layers
+            # Always have one more mixing weight, for lexical layer.
+            self.scalar_mix = scalar_mix.ScalarMix(self.max_layer + 1, do_layer_norm=False)
+
+        self.num_layers = self.model.config.num_hidden_layers
+        if args.pytorch_transformers_max_layer >= 0:
+            self.max_layer = args.pytorch_transformers_max_layer
+            assert self.max_layer <= self.num_layers
+        else:
+            self.max_layer = self.num_layers
+
+    def prepare_output(self, output_seq, lex_seq, hidden_states):
+        all_layers = [lex_seq] + list(hidden_states)
+        all_layers = all_layers[: self.max_layer + 1]
+
+        if self.embeddings_mode in ["none", "top"]:
+            h = all_layers[-1]
+        elif self.embeddings_mode == "only":
+            h = all_layers[0]
+        elif self.embeddings_mode == "cat":
+            h = torch.cat([all_layers[-1], all_layers[0]], dim=2)
+        elif self.embeddings_mode == "mix":
+            h = self.scalar_mix(all_layers, mask=mask)
+        else:
+            raise NotImplementedError(f"embeddings_mode={self.embeddings_mode}" " not supported.")
+
+        # <float32> [batch_size, var_seq_len, output_dim]
+        return h
+
+    def get_output_dim(self):
+        if self.embeddings_mode == "cat":
+            return 2 * self.model.config.hidden_size
+        else:
+            return self.model.config.hidden_size
+
+    def get_seg_ids(self, token_ids):
+        """ Dynamically build the segment IDs for a concatenated pair of sentences
+        Searches for index _sep_id in the tensor. Supports BERT or XLNet-style padding.
+
+        args:
+            token_ids (torch.LongTensor): batch of token IDs
+
+        returns:
+            seg_ids (torch.LongTensor): batch of segment IDs
+
+        example:
+        > sents = ["[CLS]", "I", "am", "a", "cat", ".", "[SEP]", "You", "like", "cats", "?", "[SEP]"]
+        > token_tensor = torch.Tensor([[vocab[w] for w in sent]]) # a tensor of token indices
+        > seg_ids = _get_seg_ids(token_tensor)
+        > assert seg_ids == torch.LongTensor([0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1])
+        """
+
+        sep_idxs = (token_ids == self._sep_id).nonzero()[:, 1]
+        seg_ids = torch.ones_like(token_ids)
+        for row, idx in zip(seg_ids, sep_idxs[::2]):
```

Comment:
Something about this seems wrong... In the single-input tasks, there's a [SEP] per input so this will still iterate when we shouldn't, I think?

Or maybe the `[:, 1]` above extracts only the second [SEP] idx per row, but then it should break on single-input cases ...?

Reply:
The default value is ones, but the for loop should run once, and fill everything in with zeros. It's not the most intuitive way to do this, but it's not my code, tests pass, and I'm pretty confident it's right.

Reply:
The tests are missing the batch size > 1 case, where the bug I think will happen?
I'll check.

Reply:
Yea, this will give incorrect segment IDs in the single-segment case :(

Reply:
lazy hack: check if the `sep_idxs.size(0) == ids.dim(0)`. If so, you don't want the `[::2]` iterator. Note this only supports 1 or 2 SEP/delimiter tokens per instance (though you can pretty easily generalize to k many by changing [::2] -> [::k].

I guess a more robust version is `k = int(sep_idxs.size(0) / ids.size(0)) ... for row, idx in zip(seg_ids, sep_idxs[::k]):`

Reply:
Fixed—seemed simplest to just write a good ol' loop-in-a-loop. I doubt this'll ever be a real performance bottleneck.